MusicCat/Players/AjaxAMP.cs

```diff
@@ -46,13 +53,102 @@ private Task<string> SendCommand(string command, Dictionary<string, string> args
+        public Task PlayFile(string filename) => Post("playfile", new Dictionary<string, string>
+        {
+	        ["filename"] = filename,
+	        ["title"] = SongList.FirstOrDefault(x => x.path == filename)?.path ?? filename
+        });
+
+	    /// <summary>
+	    /// Tells the music player to play the song with the specific id
+	    /// </summary>
+	    /// <param name="id">Id of the song to play</param>
+	    /// <returns></returns>
+	    public Task PlayID(string id) => Post("playid", new Dictionary<string, string>
+	    {
+		    ["filename"] = SongList.First(x => x.id == id).path,
+		    ["title"] = SongList.First(x => x.id == id).title
+	    });
 
         public Task SetPosition(float percent) => Post("setposition", new Dictionary<string, string> { ["pos"] = percent.ToString() });
 
         public Task SetVolume(float level) => Post("setvolume", new Dictionary<string, string> { ["level"] = level.ToString() });
 
         public Task Stop() => Post("stop");
+
+#pragma warning disable 1998
+		public async Task<List<(string name, string game, string id, float match, SongType[] types)>> Search(string[] keywords,
+		    string requiredTag = null,
+		    float cutoff = 0.3f)
+	    {
+		    var results = new List<(string name, string game, string id, float match, SongType[] types)>();
+
+			Console.WriteLine(keywords.Length);
+			Console.WriteLine(string.Join(", ", keywords));
+		    foreach (Song song in SongList)
+		    {
+				if (requiredTag != null)
+					if (song.tags == null || !song.tags.Contains(requiredTag))
+						continue;
+
+			    string[] haystack = song.title.ToLowerInvariant().Split(' ');
+			    string[] haystack2 = MetadataList.FirstOrDefault(x => x.songs.Contains(song))?.title.ToLowerInvariant()
+				    .Split(' ');
+
+			    float ratio = 0;
+			    foreach (string keyword in keywords)
+			    {
+				    string keyword2 = keyword.ToLowerInvariant();
+
+				    float subratio1 = haystack.Select(word => LevenshteinRatio(keyword2, word)).Concat(new float[] {0}).Max();
+				    float subratio2 = haystack2?.Select(word => LevenshteinRatio(keyword2, word))
+					    .Concat(new float[] {0}).Max() ?? 0;
+
+				    float subratio = (float)Math.Max(subratio1, subratio2 * 0.8);
+				    if (subratio > 0.7)
+					    ratio += subratio;
+			    }
+
+			    ratio /= keywords.Length;
+
+			    if (ratio > cutoff)
+				    results.Add((song.title, MetadataList.FirstOrDefault(x => x.songs.Contains(song))?.title
+					    .ToLowerInvariant(), song.id, ratio, song.types));
+		    }
+			Console.WriteLine(results.Count);
+		    return results.OrderByDescending(x => x.match).Take(5).ToList();
+	    }
+#pragma warning restore 1998
+
+		//https://social.technet.microsoft.com/wiki/contents/articles/26805.c-calculating-percentage-similarity-of-2-strings.aspx
+		private static float LevenshteinRatio(string source, string target)
```

Comment:
This function should live in its own file. Heck, you could even make it an extension method to string.